### PR TITLE
Rework minion return_retry_timer

### LINKED ISSUE #28577: Increase in master CPU usage after upgrading to 2015.8
After stressing 2015.8 I ran into a few issues with returns not getting back as expected. After digging into what I thought was a zmq bug, it seems that the culprit might be #27286

That change basically makes returns go from a 60s timeout to a 1-4s timeout. This incurs a lot of change in how this handles especially at scale. Before if the master was innundated with requests it had 60s to handle the response before the minion resends. With this patch we've made it so the minions can give up as early as 1s-- which for a loaded master can be quite a problem. This will not only incur retries on the minion, this also means that the master will potentially have to handle N returns for the same minion because it wasn't "fast enough" for the minion.

To get the old behavior you'd have to set `return_retry_timer` to 60 and `return_retry_random` to false. So I think we can make a more compatible (read: less intrusive) change to get comparable results.

With that in mind I'm thinking we should change a couple of things-- so here are my thoughts:

(1) longer default timeout
A range of 1-4 is probably a bit optimistic for cluster of size (at least as of today). Maybe a range more like 3-5? Just trying to increase that lower number.

(2) Configurable low-end
Right not the 1 at the bottom of the range is hard coded-- meaning even if i set the timeout to 60s Its still a spread of 1-60, where I may want to configure 30-60 or something similar.

(3) configurable retries
Since we're in here making some config options it might be nice to add an option to set the number of retries a minion will attempt for a pub return. Right now it uses the channel default (3), so maybe we add an option for that?

IMO 1 and 2 are more or less required but 3 is a nice-to-have. Any thoughts? If that all sounds well and good I can put together a PR with the necessary changes.

cc: @terminalmage

### PR DESCRIPTION
**Pull request by jacksontj**
This feature originally would randomize between 1 -> return_retry_timer, which meant you could always had the possibility of haing a return retry with a timeout of 1s. This change maxes both ends of the range configurable (similar to how acceptance_wait_time is configured). In addition I've added coverage in the minion configuration documentation.

Cleanup of #27286
Fixes #28577

### CODE REVIEW

File: salt/minion.py [hunk 1]
```diff
@@ -908,20 +908,19 @@ def _return_retry_timer(self):
...
             except ValueError:
                 # Catch wiseguys using negative integers here
                 log.error(
-                    'Invalid value ({0}) for return_retry_timer, must be a '
-                    'positive integer'.format(self.opts['return_retry_timer'])
+                    'Invalid value (return_retry_timer: {0} or return_retry_timer_max: {1})'
+                    'must be a positive integer'.format(self.opts['return_retry_timer'])
                 )
-                log.debug(msg.format(DEFAULT_MINION_OPTS['return_retry_timer']))
+                log.debug(msg.format(DEFAULT_MINION_OPTS['return_retry_timer'])
```
Comment (cachedout) [inline salt/minion.py]: Missing paren?

File: salt/minion.py [hunk 2]
```diff
@@ -908,20 +908,19 @@ def _return_retry_timer(self):
...
+                random_retry = randint(self.opts['return_retry_timer'], self.opts['return_retry_timer_max'])
+                log.debug(msg.format(random_retry) + ' (randomized)')
+                return random_retry
             except ValueError:
                 # Catch wiseguys using negative integers here
                 log.error(
-                    'Invalid value ({0}) for return_retry_timer, must be a '
-                    'positive integer'.format(self.opts['return_retry_timer'])
+                    'Invalid value (return_retry_timer: {0} or return_retry_timer_max: {1})'
+                    'must be a positive integer'.format(self.opts['return_retry_timer'])
```
Comment (cachedout) [inline salt/minion.py]: Missing second format arg for `self.opts['return_retry_timer_max']`?

**Comment by cachedout** [comment]
Quite honestly, I have no idea how 6cef894 got into this PR.

@twangboy Do you have any idea what happened here?

@jacksontj Many apologies, but would you mind either rebasing this or closing and opening another PR? Sorry about the noise.

**Comment by jacksontj** [comment]
No worries, that should do it :)

**Comment by cachedout** [comment]
@jacksontj Super. Thanks. :]

**Comment by mimianddaniel** [comment]
@cachedout was this ever merged to 2015.8 ? A bit confusing since the report initally mentioned of 2015.8 and seems like it was merged into 2015.5 branch only?

**Comment by jacksontj** [comment]
Supported branches should be regularly merged forward. So, it should be picked up with the regular merge forward (to 2015.8 and develop).

**Comment by mimianddaniel** [comment]
yep see it in 8.2

**Comment by mimianddaniel** [comment]
@jacksontj : to mimic the same behaviour as before, setting return_retry_timeout: 60 and return_retry_timer_max: 0 should do the trick? Even after setting this way, i am seeing high cpu load on the master plus, missing returns. on 2015.8.2

**Comment by jacksontj** [comment]
@mimianddaniel it is completely possible this is not a complete fix. A large change I noticed on our upgrade was the number of job returns spiked without the number of publishes increasing-- which this fixes. 

What version of 2015.8 are you running? There was another set of patches (https://github.com/saltstack/salt/pull/28410 and https://github.com/saltstack/salt/pull/28576) which re-added the missing retries to 2105.8-- which may account for your missing returns. 

As for increased CPU usage, we haven't re-rolled 2015.8 to prod yet so I don't have info on CPU usage info there :( If you have any ideas from logs/perf or whatnot I can take a look though.

**Comment by mimianddaniel** [comment]
packaged up  2015.8.2 tag + following 2015.8 branch as of yesterday and it shows two of the commits you mentioned above.  I think the missing returns are result of events that were delivered to syndics from master with huge delay.  We have multiple syndics (with about ~1200 minions per) and most of the syndics started showing delays in events from the master after upgrading to 2015.8.2 + above return timeout settings.  

As for the CPU/load,  2015.8.2 branch seems to consume more resource than current branch i am using 2015.5.5. very last peak was tested with 2015.5.5 and other peaks are with 2015.8.2
return
load
cpu

**Pull request merged**